bftengine/src/bftengine/ReplicaImp.cpp

```diff
@@ -1841,13 +1842,15 @@ void ReplicaImp::onMessage<CheckpointMsg>(CheckpointMsg *msg) {
 
   bool askForStateTransfer = false;
 
-  if (msgIsStable && msgSeqNum > lastExecutedSeqNum) {
+  if (msgIsStable && (msgSeqNum > lastExecutedSeqNum || msgEpochNum >= getSelfEpochNumber())) {
     auto pos = tableOfStableCheckpoints.find(msgGenReplicaId);
-    if (pos == tableOfStableCheckpoints.end() || pos->second->seqNumber() <= msgSeqNum) {
+    if (pos == tableOfStableCheckpoints.end() || pos->second->seqNumber() <= msgSeqNum ||
+        msgEpochNum >= getSelfEpochNumber()) {
```

Comment:
Just to understand
1. The checkpoint message is the only one allowed to be received if the message epoch >= replica epoch, right?
2. We still start state transfer only based on sequence number?

Reply:
yes (1) is True, i the reason i did not add this is valid method because it is called for RO replica as well and there is no self epoch number for RO replica